### PR TITLE
Implement C Level Timedelta ISO Function; fix JSON usage

### LINKED ISSUE #28256: Timedelta in to_json object array and ISO dates not handled properly
Intertwined with #15137 though a slight different issue.

`date_format="iso"` has different behavior for Timedeltas depending on whether or not the Timedelta is in a DTA or an object array. To illustrate:

```python
# Wrong format ref 15137, but at least tries to do some formatting
>>> pd.DataFrame([[pd.Timedelta("1D")]]).to_json(date_format="iso")
'{"0":{"0":"1970-01-02T00:00:00.000Z"}}'

# Object array has no formatting
>>> pd.DataFrame([[pd.Timedelta("1D")]]).astype(object).to_json(date_format="iso")
'{"0":{"0":86400000}}'
```

By contrast the same issue does not appear with datetimes
```python
>>> pd.DataFrame([[pd.Timestamp(1)]]).to_json(date_format="iso")
'{"0":{"0":"1970-01-01T00:00:00.000Z"}}'

# Below still formats as iso in spite of being object array
>>> pd.DataFrame([[pd.Timestamp(1)]]).astype(object).to_json(date_format="iso")
'{"0":{"0":"1970-01-01T00:00:00.000Z"}}'
```

### PR DESCRIPTION
**Pull request by WillAyd**
- [X] closes #28256
- [X] tests added / passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

similar to @cbertinato great work in #28595 but done deeper in the C extension

### CODE REVIEW

File: pandas/_libs/src/ujson/python/objToJSON.c [hunk 1]
```diff
@@ -407,6 +406,30 @@ static char *int64ToIso(int64_t value, NPY_DATETIMEUNIT base, size_t *len) {
     return result;
 }
 
+/* Converts the int64_t representation of a duration to ISO; mutates len */
+static char *int64ToIsoDuration(int64_t value, size_t *len) {
```
Comment (WillAyd) [inline pandas/_libs/src/ujson/python/objToJSON.c]: I think after this PR going to split these conversion functions off into a separate module as the JSON one is getting rather big

File: pandas/_libs/src/ujson/python/objToJSON.c [hunk 2]
```diff
@@ -1615,6 +1649,11 @@ char **NpyArr_encodeLabels(PyArrayObject *labels, PyObjectEncoder *enc,
         ret[i] = PyObject_Malloc(len + 1);
         memcpy(ret[i], cLabel, len + 1);
 
+        if (is_datetimelike) {
```
Comment (WillAyd) [inline pandas/_libs/src/ujson/python/objToJSON.c]: This fixes a memory leak in the current implementation where any datetimelike values used in the index would leak their respective character representation out; will add a whatsnew once the file is created

**Comment by WillAyd** [comment]
Current benchmarks:

```sh
       before           after         ratio
     [447a3b00]       [4a94f159]
     <master>         <json-timedelta>
+         152±3ms         209±10ms     1.38  io.json.ToJSONISO.time_iso_format('values')
+         168±3ms         213±10ms     1.27  io.json.ToJSONISO.time_iso_format('records')
+         188±4ms         235±10ms     1.25  io.json.ToJSONISO.time_iso_format('split')
+         195±4ms         238±10ms     1.22  io.json.ToJSONISO.time_iso_format('columns')
-        103±10ms         79.2±6ms     0.77  io.json.ToJSON.time_to_json('values', 'df_td_int_ts')
-        137±20ms         95.9±1ms     0.70  io.json.ToJSON.time_to_json('values', 'df_int_float_str')
```

So the time_iso_format checks are taking a little bit longer but note that they were broken before, i.e. on master you get this:

```python
>>> pd.Series([pd.Timedelta(days=1)], index=[pd.Timedelta(days=1)]).to_json(date_format="iso")
'{"P1DT0H0M0S":"1970-01-02T00:00:00.000Z"}'
```
Where we are writing as dates and not durations.

I think just a matter of the timedelta -> iso function not being as fast (using sprintf instead of totally hand-coded) but maybe OK for now

File: pandas/tests/io/json/test_pandas.py [hunk 3]
```diff
@@ -810,6 +811,29 @@ def test_convert_dates(self):
...
+            data.append("a")
+
+        ser = pd.Series(data, index=data)
+        result = ser.to_json(date_format=date_format)
+
+        if date_format == "epoch":
+            expected = '{"1577836800000":1577836800000,"null":null}'
+        else:
+            expected = (
+                '{"2020-01-01T00:00:00.000Z":"2020-01-01T00:00:00.000Z"' ',"null":null}'
```
Comment (WillAyd) [inline pandas/tests/io/json/test_pandas.py]: I think CI failing on this is a false positive - @MomIsBestFriend any thoughts?

Comment (MomIsBestFriend) [reply pandas/tests/io/json/test_pandas.py]: @WillAyd I don't think the CI have a false positive. but I do want to understand what you're trying to do, are you trying ```expected``` to be a tuple, or just one string? if the answer is tuple I think you've missed a coma and the value of ```expected``` should be:

```python
expected = (
'{"2020-01-01T00:00:00.000Z":"2020-01-01T00:00:00.000Z"', ',"null":null}'
)
```

and if the value of ```expected``` spouse to be just a regular string, please just concat the two strings, so the value of ```expected``` will be:

```python
expected = (
                '{"2020-01-01T00:00:00.000Z":"2020-01-01T00:00:00.000Z" ,"null":null}'
)
```

---

And if neither of these is the answer, please explain to me what the type of ```expected``` should be, and just disable the CI. I will work on fixing it.

Comment (WillAyd) [reply pandas/tests/io/json/test_pandas.py]: Ah i see now. Thanks for confirming

File: pandas/_libs/src/ujson/python/objToJSON.c (hunk 1 above)
Comment (WillAyd) [reply pandas/_libs/src/ujson/python/objToJSON.c]: The other PR is #31057 should probably be merged first

File: pandas/_libs/tslibs/src/datetime/np_datetime_strings.c [hunk 4]
```diff
@@ -905,3 +905,37 @@ int make_iso_8601_datetime(npy_datetimestruct *dts, char *outstr, int outlen,
                  outlen);
     return -1;
 }
+
+
+int make_iso_8601_timedelta(pandas_timedeltastruct *tds,
```
Comment (WillAyd) [inline pandas/_libs/tslibs/src/datetime/np_datetime_strings.c]: Note that we do also have a isoformat function for time deltas defined here:

https://github.com/pandas-dev/pandas/blob/cdffa43af34aa832105dd3787eec20ea0222af54/pandas/_libs/tslibs/timedeltas.pyx#L1097

But I think the path to get that to work in the JSON module would be way more complicated than implementing in C as we do here (and less consistent with other datetime handling). 

Maybe as a follow up we want to use this C function in the Timedelta class

**Comment by jbrockmendel** [comment]
pls rebase

**Comment by jbrockmendel** [comment]
LGTM cc @jreback

**Comment by jbrockmendel** [comment]
needs rebase, other good to go i think

**Comment by jreback** [comment]
completely closes #28256. ?

can you check the perf again.

**Comment by WillAyd** [comment]
Yea closes; added some more test coverage as well.

Here is an updated benchmark:

```sh
       before           after         ratio
     [402c5cdd]       [ef08ad6c]
     <json-test-cleanup~1^2>       <json-timedelta>
+        143±20ms         225±10ms     1.57  io.json.ToJSONISO.time_iso_format('values')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

**Comment by jreback** [comment]
this is slower?

**Comment by WillAyd** [comment]
Slower but correct

**Comment by jbrockmendel** [comment]
> Slower but correct

worth the tradeoff every time.

**Comment by jreback** [comment]
thanks, correct wins every time.

**Pull request merged**